composeApp/src/commonMain/kotlin/sp/bvantur/tasky/register/presentation/RegisterViewModel.kt

```diff
@@ -0,0 +1,120 @@
+package sp.bvantur.tasky.register.presentation
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.debounce
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.launch
+import sp.bvantur.tasky.core.DispatcherProvider
+import sp.bvantur.tasky.core.domain.ValidateEmailUseCase
+import sp.bvantur.tasky.core.domain.ValidateNameUseCase
+import sp.bvantur.tasky.core.domain.ValidatePasswordUseCase
+import sp.bvantur.tasky.core.presentation.ViewModelUserActionHandler
+import sp.bvantur.tasky.core.presentation.ViewModelViewStateHandler
+import sp.bvantur.tasky.core.presentation.ViewModelViewStateHandlerImpl
+import kotlin.time.Duration.Companion.milliseconds
+
+typealias OnRegisterUserAction = (RegisterUserAction) -> Unit
+
+@OptIn(FlowPreview::class)
+class RegisterViewModel(
+    private val dispatcherProvider: DispatcherProvider,
+    private val validateNameUseCase: ValidateNameUseCase,
+    private val validateEmailUseCase: ValidateEmailUseCase,
+    private val validatePasswordUseCase: ValidatePasswordUseCase
+) : ViewModel(),
+    ViewModelUserActionHandler<RegisterUserAction>,
+    ViewModelViewStateHandler<RegisterViewState> by ViewModelViewStateHandlerImpl(
+        RegisterViewState(),
+        dispatcherProvider
+    ) {
+
+    private val textDebounceFlow = MutableSharedFlow<Pair<String, suspend (String) -> Unit>>()
+
+    init {
+        textDebounceFlow
+            .debounce(500.milliseconds)
+            .onEach { (text, action) ->
+                action(text)
+            }
+            .launchIn(CoroutineScope(dispatcherProvider.default))
```

Comment:
Could it make more sense here to launch the collection of the `Flow` in the `viewModelScope`? This will automatically cancel all coroutines launched in this scope when the ViewModel is destroyed since they are tied to the ViewModel’s lifecycle.

You’ve created a custom `CoroutineScope` which you have no control over and it's not lifecycle-aware, meaning you have the potential of memory leaks if the scope outlives the ViewModel.

Say you performed a network call to search for records matching your search text, you’d want a way to cancel the current search if the user changed their search query before the previous network call finishes because you no longer care about that result.

Reply:
Good catch! 👍 I removed that code but this is a valid point to take with me. Thanks! :)